File: packages/platos-client-py/platos_client/apis/trigger.py
"""Background-operations (BGO) surface for platos-client.

Theme I.6 parity with TS I.2; renamed under Theme BGO. Exposed on
``PlatosClient`` as both ``client.bgo`` (canonical) and ``client.trigger``
(deprecated alias, removed in the next major). Both point at the same
``TriggerApi`` instance. See ``docs/BGO_RENAME.md``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from platos_client.client import PlatosClient, PlatosScope


class TriggerApi:
    def __init__(self, client: "PlatosClient") -> None:
        self.tasks = TriggerTasksApi(client)
        self.runs = TriggerRunsApi(client)
        self.schedules = TriggerSchedulesApi(client)
        self.batches = TriggerBatchesApi(client)
        self._client = client

    async def raw(
        self,
        path: str,
        *,
        method: str = "GET",
        body: Any = None,
        scope: "PlatosScope | None" = None,
    ) -> Any:
        return await self._client._request(method, path, scope=scope, body=body)


class TriggerTasksApi:
    def __init__(self, client: "PlatosClient") -> None:
        self._client = client

    async def list(self, scope: "PlatosScope | None" = None) -> list[dict[str, Any]]:
        res = await self._client._request(
            "GET", "/api/v1/agent/trigger/tasks", scope=scope
        )
        return list((res or {}).get("tasks", []))

    async def trigger(
        self,
        task_id: str,
        payload: Any,
        *,
        options: dict[str, Any] | None = None,
        scope: "PlatosScope | None" = None,
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST",
            f"/api/v1/agent/trigger/tasks/{task_id}/trigger",
            scope=scope,
            body={"payload": payload, "options": options or {}},
        )


class TriggerRunsApi:
    def __init__(self, client: "PlatosClient") -> None:
        self._client = client

    async def list(
        self,
        *,
        task_id: str | None = None,
        status: str | None = None,
        limit: int | None = None,
        scope: "PlatosScope | None" = None,
    ) -> list[dict[str, Any]]:
        qs = []
        if task_id:
            qs.append(f"taskId={task_id}")
        if status:
            qs.append(f"status={status}")
        if limit:
            qs.append(f"limit={limit}")
        suffix = ("?" + "&".join(qs)) if qs else ""
        res = await self._client._request(
            "GET", f"/api/v1/agent/trigger/runs{suffix}", scope=scope
        )
        return list((res or {}).get("runs", []))

    async def get(
        self, run_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any]:
        return await self._client._request(
            "GET", f"/api/v1/agent/trigger/runs/{run_id}", scope=scope
        )

    async def cancel(
        self, run_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/api/v1/agent/trigger/runs/{run_id}/cancel", scope=scope
        )

    async def replay(
        self, run_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST", f"/api/v1/agent/trigger/runs/{run_id}/replay", scope=scope
        )


class TriggerSchedulesApi:
    def __init__(self, client: "PlatosClient") -> None:
        self._client = client

    async def list(self, scope: "PlatosScope | None" = None) -> list[dict[str, Any]]:
        res = await self._client._request(
            "GET", "/api/v1/agent/trigger/schedules", scope=scope
        )
        return list((res or {}).get("schedules", []))

    async def create(
        self,
        *,
        task_id: str,
        cron: str,
        timezone: str | None = None,
        payload: Any = None,
        scope: "PlatosScope | None" = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"taskId": task_id, "cron": cron}
        if timezone:
            body["timezone"] = timezone
        if payload is not None:
            body["payload"] = payload
        return await self._client._request(
            "POST", "/api/v1/agent/trigger/schedules", scope=scope, body=body
        )

    async def activate(
        self, schedule_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST",
            f"/api/v1/agent/trigger/schedules/{schedule_id}/activate",
            scope=scope,
        )

    async def deactivate(
        self, schedule_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST",
            f"/api/v1/agent/trigger/schedules/{schedule_id}/deactivate",
            scope=scope,
        )

    async def delete(
        self, schedule_id: str, scope: "PlatosScope | None" = None
    ) -> None:
        await self._client._request(
            "DELETE", f"/api/v1/agent/trigger/schedules/{schedule_id}", scope=scope
        )


class TriggerBatchesApi:
    def __init__(self, client: "PlatosClient") -> None:
        self._client = client

    async def trigger(
        self,
        task_id: str,
        payloads: list[Any],
        *,
        options: dict[str, Any] | None = None,
        scope: "PlatosScope | None" = None,
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST",
            f"/api/v1/agent/trigger/batches/{task_id}",
            scope=scope,
            body={"payloads": payloads, "options": options or {}},
        )

    async def get(
        self, batch_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any]:
        return await self._client._request(
            "GET", f"/api/v1/agent/trigger/batches/{batch_id}", scope=scope
        )

File: packages/platos-client/src/apis/trigger.ts
/**
 * @platosdev/client — background-operations (BGO) surface (Theme I.2; renamed
 * under Theme BGO).
 *
 * Mirrors the agent's B.5 meta-tools as first-class SDK calls so
 * integrators don't need two clients. Every method reuses
 * `PlatosClient._fetch` (retry, 401-refresh, typed errors) and the same
 * auth mode the caller configured on the client.
 *
 * Exposed on `PlatosClient` as both `client.bgo` (canonical, Theme BGO)
 * and `client.trigger` (deprecated alias — removed in the next major).
 * Both point at the same `TriggerApi` instance; the internal class name
 * is retained as `TriggerApi` because it wraps the underlying trigger.dev
 * engine. See `docs/BGO_RENAME.md`.
 *
 * Convention:
 *   - `bgo.tasks.*`    → background-operation catalog (list / trigger)
 *   - `bgo.runs.*`     → individual run lifecycle (get / cancel /
 *                        replay / list)
 *   - `bgo.schedules.*`→ schedule CRUD
 *   - `bgo.batches.*`  → batch spawns
 *
 * For ops that the agent service already mirrors (e.g. `list_bgos` a.k.a.
 * `list_tasks`, `list_runs`) we target the agent `/api/v1/agent/trigger/*`
 * shim which enforces scope filtering; the pure trigger.dev API under
 * `/api/v1/*` on the webapp is still accessible directly via `bgo.raw()`.
 */

import type { PlatosClient } from "../client.js";
import type { PlatosScope } from "../types.js";

export interface TriggerTaskCatalogEntry {
  id: string;
  name?: string;
  filePath?: string;
  exportName?: string;
  [extra: string]: unknown;
}

export interface TriggerRunSummary {
  id: string;
  taskId: string;
  status: string;
  createdAt: string;
  startedAt?: string | null;
  finishedAt?: string | null;
  error?: unknown;
  output?: unknown;
  metadata?: Record<string, unknown> | null;
  [extra: string]: unknown;
}

export interface TriggerScheduleSummary {
  id: string;
  taskId: string;
  cron: string;
  timezone?: string | null;
  active: boolean;
  lastRunAt?: string | null;
  nextRunAt?: string | null;
  [extra: string]: unknown;
}

export interface TriggerTaskOptions {
  idempotencyKey?: string;
  metadata?: Record<string, unknown>;
  delay?: string | Date;
  ttl?: string;
  tags?: string[];
  [extra: string]: unknown;
}

export interface TriggerHandle {
  id: string;
  publicAccessToken?: string;
  [extra: string]: unknown;
}

export class TriggerApi {
  readonly tasks: TriggerTasksApi;
  readonly runs: TriggerRunsApi;
  readonly schedules: TriggerSchedulesApi;
  readonly batches: TriggerBatchesApi;

  constructor(private readonly client: PlatosClient) {
    this.tasks = new TriggerTasksApi(client);
    this.runs = new TriggerRunsApi(client);
    this.schedules = new TriggerSchedulesApi(client);
    this.batches = new TriggerBatchesApi(client);
  }

  /**
   * Escape hatch — hit any trigger.dev webapp endpoint directly. Scope
   * headers are attached automatically; pass `body` / method as usual.
   *
   *   const res = await client.bgo.raw("/api/v1/runs?limit=10");
   *   // or (deprecated alias): client.trigger.raw(...)
   */
  async raw<T = unknown>(
    path: string,
    init: RequestInit = {},
    scope?: PlatosScope,
  ): Promise<T> {
    return this.client._fetch<T>(path, init, scope);
  }
}

class TriggerTasksApi {
  constructor(private readonly client: PlatosClient) {}

  /** List background operations in the caller's scope (mirrors the `list_bgos` meta-tool — formerly `list_tasks`). */
  async list(scope?: PlatosScope): Promise<TriggerTaskCatalogEntry[]> {
    const res = await this.client._fetch<{ tasks: TriggerTaskCatalogEntry[] }>(
      "/api/v1/agent/trigger/tasks",
      { method: "GET" },
      scope,
    );
    return res?.tasks ?? [];
  }

  /**
   * Trigger a task by id. Returns a handle containing the `runId` —
   * feed that to `runs.get()` or subscribe via a `threads.send` stream
   * for realtime updates.
   */
  async trigger<TPayload = unknown>(
    taskId: string,
    payload: TPayload,
    options: TriggerTaskOptions = {},
    scope?: PlatosScope,
  ): Promise<TriggerHandle> {
    return this.client._fetch<TriggerHandle>(
      `/api/v1/agent/trigger/tasks/${encodeURIComponent(taskId)}/trigger`,
      {
        method: "POST",
        body: JSON.stringify({ payload, options }),
      },
      scope,
    );
  }
}

class TriggerRunsApi {
  constructor(private readonly client: PlatosClient) {}

  /** Recent runs for the caller's scope. Mirrors the `list_runs` meta-tool. */
  async list(
    query: { taskId?: string; status?: string; limit?: number } = {},
    scope?: PlatosScope,
  ): Promise<TriggerRunSummary[]> {
    const qs = new URLSearchParams();
    if (query.taskId) qs.set("taskId", query.taskId);
    if (query.status) qs.set("status", query.status);
    if (query.limit) qs.set("limit", String(query.limit));
    const suffix = qs.toString() ? `?${qs.toString()}` : "";
    const res = await this.client._fetch<{ runs: TriggerRunSummary[] }>(
      `/api/v1/agent/trigger/runs${suffix}`,
      { method: "GET" },
      scope,
    );
    return res?.runs ?? [];
  }

  /** Fetch a run's status + output + recent logs. */
  async get(runId: string, scope?: PlatosScope): Promise<TriggerRunSummary> {
    return this.client._fetch<TriggerRunSummary>(
      `/api/v1/agent/trigger/runs/${encodeURIComponent(runId)}`,
      { method: "GET" },
      scope,
    );
  }

  /** Cancel a running run. Mirrors `cancel_run` meta-tool — approval gated. */
  async cancel(runId: string, scope?: PlatosScope): Promise<{ status: string }> {
    return this.client._fetch<{ status: string }>(
      `/api/v1/agent/trigger/runs/${encodeURIComponent(runId)}/cancel`,
      { method: "POST" },
      scope,
    );
  }

  /** Replay a finished / failed run. */
  async replay(runId: string, scope?: PlatosScope): Promise<TriggerHandle> {
    return this.client._fetch<TriggerHandle>(
      `/api/v1/agent/trigger/runs/${encodeURIComponent(runId)}/replay`,
      { method: "POST" },
      scope,
    );
  }
}

class TriggerSchedulesApi {
  constructor(private readonly client: PlatosClient) {}

  async list(scope?: PlatosScope): Promise<TriggerScheduleSummary[]> {
    const res = await this.client._fetch<{ schedules: TriggerScheduleSummary[] }>(
      "/api/v1/agent/trigger/schedules",
      { method: "GET" },
      scope,
    );
    return res?.schedules ?? [];
  }

  async create(
    body: { taskId: string; cron: string; timezone?: string; payload?: unknown },
    scope?: PlatosScope,
  ): Promise<TriggerScheduleSummary> {
    return this.client._fetch<TriggerScheduleSummary>(
      "/api/v1/agent/trigger/schedules",
      { method: "POST", body: JSON.stringify(body) },
      scope,
    );
  }

  async activate(scheduleId: string, scope?: PlatosScope): Promise<TriggerScheduleSummary> {
    return this.client._fetch<TriggerScheduleSummary>(
      `/api/v1/agent/trigger/schedules/${encodeURIComponent(scheduleId)}/activate`,
      { method: "POST" },
      scope,
    );
  }

  async deactivate(scheduleId: string, scope?: PlatosScope): Promise<TriggerScheduleSummary> {
    return this.client._fetch<TriggerScheduleSummary>(
      `/api/v1/agent/trigger/schedules/${encodeURIComponent(scheduleId)}/deactivate`,
      { method: "POST" },
      scope,
    );
  }

  async delete(scheduleId: string, scope?: PlatosScope): Promise<void> {
    await this.client._fetch<void>(
      `/api/v1/agent/trigger/schedules/${encodeURIComponent(scheduleId)}`,
      { method: "DELETE" },
      scope,
    );
  }
}

class TriggerBatchesApi {
  constructor(private readonly client: PlatosClient) {}

  /** Fire N tasks in parallel via a single call. Mirrors `spawn_batch`. */
  async trigger<TPayload = unknown>(
    taskId: string,
    payloads: TPayload[],
    options: TriggerTaskOptions = {},
    scope?: PlatosScope,
  ): Promise<{ batchId: string; runIds: string[] }> {
    return this.client._fetch<{ batchId: string; runIds: string[] }>(
      `/api/v1/agent/trigger/batches/${encodeURIComponent(taskId)}`,
      {
        method: "POST",
        body: JSON.stringify({ payloads, options }),
      },
      scope,
    );
  }

  async get(batchId: string, scope?: PlatosScope): Promise<{ id: string; status: string; runIds: string[] }> {
    return this.client._fetch<{ id: string; status: string; runIds: string[] }>(
      `/api/v1/agent/trigger/batches/${encodeURIComponent(batchId)}`,
      { method: "GET" },
      scope,
    );
  }
}
